### ISSUE TITLE
Legacy EC public key encoder/decoder function issues

### ISSUE THREAD

**Issue by kaduk** [openssl/openssl#14258]
There are some issues with the `i2d_PublicKey()` and `d2i_PublicKey()` functions.

- Calling i2d_PublicKey() on any provided EC public key ends up in i2d_provided() trying to produce "type-specific" output, but none of the candidates found in the encoder context end up claiming to match that, so we fail to produce any output.
- After forcing EC keys to use the "ec" output structure format, the encoder call succeeds, but returns a bogus length (roughly `INT_MAX ` minus the actual length
- It should be possible to use `d2i_PublicKey()` on the encoded output, but `d2i_PublicKey()` for EC keys has the unfortunate API requirement that an input `EVP_PKEY *` is supplied that contains an `EC_KEY` of the proper `EC_GROUP` needed to decode the public key.  I was unable to find a non-deprecated API that allows such a skeleton `EVP_PKEY` to be produced.  (Maybe I missed one, but the obvious ones involving `EVP_PKEY_assign_EC_KEY()` are deprecated.)  Thus, as far as I can tell, `d2i_PublicKey()` itself is effectively deprecated for use with EC keys.

**Comment by t8m** [comment]
You can use EVP_PKEY_fromdata() to create EVP_PKEY with appropriate EC parameters.
```
OSSL_PARAM params[2];

params[0] = OSSL_PARAM_construct_utf8_string(OSSL_PKEY_PARAM_GROUP_NAME,
                                                                                 curve_name, 0);
params[1] = OSSL_PARAM_construct_end();
pctx = EVP_PKEY_CTX_new_from_name(NULL, "EC", NULL);
EVP_PKEY_fromdata_init(pctx);
EVP_PKEY_fromdata(pctx, &pkey,
                                    OSSL_KEYMGMT_SELECT_DOMAIN_PARAMETERS,
                                    params);
```

**Comment by kaduk** [comment]
> You can use EVP_PKEY_fromdata() to create EVP_PKEY with appropriate EC parameters.

Ah, thanks for the pointer.  I knew about parameters-only keys, and I knew about `EVP_PKEY_fromdata_init()` (it's what the actual application is using now, to just build the key directly), but I didn't make the connection of using the one to make the other.

**Comment by levitte** [comment]
Fixes in #14291.  I have *not* done anything about d2i_PublicKey(), however.

**Comment by kaduk** [comment]
> Fixes in #14291. I have _not_ done anything about d2i_PublicKey(), however.

Thanks for putting together the fixes!

IIUC `d2i_PublicKey()` will just give you a legacy key, which seems fine.

**Comment by levitte** [comment]
I hope I haven't missed anything...  I'm too tired to look, but am running tests and will have another look tomorrow morning
